gripper_action_controller/include/gripper_action_controller/gripper_action_controller_impl.h

```diff
@@ -0,0 +1,380 @@
+    RCLCPP_ERROR(node_->get_logger(), "Expected 1 velocity state interfaces");
+    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::
+    CallbackReturn::ERROR;
+  }
+  if (velocity_state_interface_it->get_name() != joint_name_) {
+    RCLCPP_ERROR_STREAM(
+        node_->get_logger(),
+        "Velocity command interface is different than joint name `"
+            << velocity_state_interface_it->get_name() << "` != `"
+            << joint_name_ << "`");
+    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::
+    CallbackReturn::ERROR;
+  }
+
+  joint_position_command_interface_ = &(*position_command_interface_it);
+  joint_position_state_interface_ = &(*position_state_interface_it);
+  joint_velocity_state_interface_ = &(*velocity_state_interface_it);
+
+  // Hardware interface adapter
+  hw_iface_adapter_.init(joint_position_command_interface_, node_);
+
+  // Command - non RT version
+  command_struct_.position_ = joint_position_state_interface_->get_value();
+  command_struct_.max_effort_ = default_max_effort_;
+
+  // Result
+  pre_alloc_result_ =
+      std::make_shared<control_msgs::action::GripperCommand::Result>();
+  pre_alloc_result_->position = command_struct_.position_;
+  pre_alloc_result_->reached_goal = false;
+  pre_alloc_result_->stalled = false;
+
+  // Action interface
+  using namespace std::placeholders;
+  action_server_ =
+      rclcpp_action::create_server<control_msgs::action::GripperCommand>(
+          node_, "~/gripper_cmd",
+          std::bind(&GripperActionController::goal_callback, this, _1, _2),
+          std::bind(&GripperActionController::cancel_callback, this, _1),
+          std::bind(&GripperActionController::accepted_callback, this, _1));
+
+  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::
+  CallbackReturn::SUCCESS;
+}
+
+template <const char *HardwareInterface>
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+GripperActionController<HardwareInterface>::on_deactivate(
+    const rclcpp_lifecycle::State &previous_state) {
+  joint_position_command_interface_ = nullptr;
+  joint_position_state_interface_ = nullptr;
+  joint_velocity_state_interface_ = nullptr;
+  release_interfaces();
+  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::
+  CallbackReturn::SUCCESS;
+}
+
+template <const char *HardwareInterface>
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+GripperActionController<HardwareInterface>::on_cleanup(
+    const rclcpp_lifecycle::State &previous_state) {
+  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::
+  CallbackReturn::SUCCESS;
+}
+
+template <const char *HardwareInterface>
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+GripperActionController<HardwareInterface>::on_error(
+    const rclcpp_lifecycle::State &previous_state) {
+  return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::
```

Comment:
I think we can do without the boilerplate

https://github.com/ros2/rclcpp/blob/master/rclcpp_lifecycle/src/node_interfaces/lifecycle_node_interface.cpp